operator: fix upgradeImages

hj-johannes-lee opened the pull request:
fixes: #1873 
Operatorhub bundle can have sha256 image tags that are put through env vars. When operator controller manager gets upgraded, its operands (plugin daemonsets) should be updated to the image in the env vars. But it has not been working properly because of wrong parsing.

Fix it to parse the image names that have sha256 tags correctly so env vars in operator can be used as intended.

mythi commented:
> For sha256 image tags, it has not been working properly.

Can you write how you expect it to work and how it works now. IIRC we decided SHA pinning is only for the env var approach.

hj-johannes-lee commented:
If image is `intel-qat-plugin@sha256:69858421d6abd38867f61844db2c3fa7275bf6c089efd547e7de98dfc0c90a1a`

Before:
`name` was `intel-qat-plugin@sha256`
and upgradeImage was not working even though envVar is properly set.

For example, let's say there is 0.30.0 version of deployed operatorhub bundle using olm (as written in the [link](https://github.com/intel/intel-device-plugins-for-kubernetes/blob/main/DEVEL.md#publish-a-new-version-of-the-intel-device-plugins-operator-to-operatorhubio)), and then let's say we deploy a qat plugin (0.30.0 version).
And then we update the bundle with the comamnd:
```
operator-sdk run bundle-upgrade intel/intel-deviceplugins-operator:0.31.0
```
Then, the qat-plugin is expected to be upgraded automatically according to the envVar, but it does not because the parsing has been wrong (`intel-qat-plugin@sha256`)

So,
I expect `name` to be `intel-qat-plugin` so that it can be updated when another sha256 is found from envVar.
And it works now like that, and tested with olm.

mythi commented:
Thanks for noticing this! TestUpgrade unit test needs updating to get this covered. Also .1 release is needed

hj-johannes-lee commented:
Yeah, I think we should have had it early... It's my mistake.
I pushed the commit for the testUpgrade fix too.

hj-johannes-lee commented on pkg/controllers/reconciler_test.go:
```diff
@@ -61,11 +66,19 @@ func TestUpgrade(test *testing.T) {
 	for i := range tests {
 		t := tests[i]
-
+		os.Setenv(strings.ReplaceAll(strings.ToUpper(filepath.Base(t.image)), "-", "_")+"_SHA", t.image+"@sha256:"+sha256)
```
Umm what would be the way to put the `flag`?
another `bool` type field after `upgrade bool`?

hj-johannes-lee replied:
I put `isEnvVarSet`. If it's not what you meant, please comment again.!

hj-johannes-lee replied:
I do not understand what you mean.

hj-johannes-lee replied:
I think I understood what you mean, and changed.
But, I am not sure if it is 'simpler' though I think it looks 'clearer'.

One thing I am not sure is if it would look better if I make a function for
`strings.ReplaceAll(strings.ToUpper(filepath.Base(name)), "-", "_") + "_SHA"` part?

tkatila replied:
I think having a function for the ReplaceAll-ToUpper combo would make it look better. Especially with the test code. Feel free to make the change.

hj-johannes-lee replied:
@tkatila Thanks for your comment.!

I was thinking that looks better, but Mikko suggested to "hardcode" that part.
I thought it makes sense and looks intuitive to readers. So, let me just.. use that.

tkatila commented on pkg/controllers/reconciler_test.go:
```diff
@@ -61,11 +67,24 @@ func TestUpgrade(test *testing.T) {
 	for i := range tests {
 		t := tests[i]
 
+		for key, value := range t.envVars {
+			os.Setenv(key, value)
+		}
```
Env variables stay from test run to test run. I think we should somehow clear the variables after the run. If more tests are added which do not use the env variables, they could get false-negative results.

hj-johannes-lee replied:
Thanks! Added `os.clearEnv()`!

tkatila commented on pkg/controllers/reconciler_test.go:
```diff
@@ -61,18 +67,33 @@ func TestUpgrade(test *testing.T) {
 			test.Errorf("expectedInitimage: %s, received: %s", t.expectedInitimage, t.initimage)
 		}
+
+		os.Clearenv()
```
This will delete everything, it should only remove the ones that were added.

hj-johannes-lee replied:
fixed to `os.UnsetEnv`

mythi commented on pkg/controllers/reconciler_test.go:
```diff
@@ -61,18 +67,35 @@ func TestUpgrade(test *testing.T) {
 			upgrade:           false,
 		},
+		{
+			envVars: map[string]string{strings.ReplaceAll(strings.ToUpper(filepath.Base(image)), "-", "_") + "_SHA": image + "@sha256:" + sha256,
+				strings.ReplaceAll(strings.ToUpper(filepath.Base(initimage)), "-", "_") + "_SHA": initimage + "@sha256:" + sha256},
```
This is rather complex for the reader to understand what this test is all about. I'd suggest to spell out the strings for readability, maybe with a different image name:
```suggestion
			envVars: map[string]string{
			        INTEL_QAT_PLUGIN_SHA: "intel/intel-qat-plugin@sha256:000000000000000000000000000000000000000000000000000000000000000a",
			        INTEL_QAT_INITIMAGE_SHA: "intel/intel-qat-initimage@sha256:000000000000000000000000000000000000000000000000000000000000000a",
				},
```

hj-johannes-lee replied:
If you see the image name, it is fixed as dsa plugin's image name. Is the qat example just an example?

mythi reviewed (changes requested):
for full coverage, we'd also need a case where `image` is same as the corresponding envVar and when the envVar is not in the pinned/sha256 format.

hj-johannes-lee commented:
@mythi 
> for full coverage, we'd also need a case where image is same as the corresponding envVar and when the envVar is not in the pinned/sha256 format.

Umm, we do not check in `reconciler.go` whether the envVar value is sha256 format or not.
In our current implementation, we always expect the envvar value for the image would be in sha256 format.
The name of the envvar itself is like `INTEL_DSA_PLUGIN_SHA`

So, do you want to improve the implementation? Or... did I misunderstand what you mean?

mythi commented:
> So, do you want to improve the implementation?

we can focus on the current functionality and make sure all scenarios are covered

hj-johannes-lee commented:
@mythi
Thanks for the comment. I think we all cover for the "current implementation."

hj-johannes-lee commented:
Fixed.!

And,, added some comments in the code to make it understandable more easily (actually for myself to understand next time.. I need to always follow the process and understand again...). If you think it's not good, let me konw.!

mythi reviewed (approved):
LGTM, thanks!
